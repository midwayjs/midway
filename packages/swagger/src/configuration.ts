import {
  ILifeCycle,
  IMidwayContainer,
  MidwayApplicationManager,
  MidwayConfigService,
  Inject,
  Configuration,
} from '@midwayjs/core';
import { SwaggerExplorer, SwaggerMiddleware } from '.';
import * as DefaultConfig from './config/config.default';

@Configuration({
  importConfigs: [
    {
      default: DefaultConfig,
    },
  ],
  namespace: 'swagger',
})
export class SwaggerConfiguration implements ILifeCycle {
  @Inject()
  applicationManager: MidwayApplicationManager;

  @Inject()
  configService: MidwayConfigService;

  async onReady(container: IMidwayContainer) {
    const apps = this.applicationManager.getApplications([
      'express',
      'koa',
      'egg',
      'faas',
    ]);

    if (apps.length) {
      let globalPrefix =
        this.configService.getConfiguration('globalPrefix') ||
        this.configService.getConfiguration('koa.globalPrefix') ||
        this.configService.getConfiguration('express.globalPrefix') ||
        this.configService.getConfiguration('egg.globalPrefix');
      const explorer = await container.getAsync(SwaggerExplorer);

      if (globalPrefix) {
        if (!/^\//.test(globalPrefix)) {
          globalPrefix = '/' + globalPrefix;
        }

        explorer.addGlobalPrefix(globalPrefix);
      }

      explorer.scanApp();

      apps.forEach(app => {
        app.useMiddleware(SwaggerMiddleware);
      });
    }
  }
}
