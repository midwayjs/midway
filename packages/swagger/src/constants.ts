export const DECORATORS_PREFIX = 'swagger';
export const DECORATORS = {
  API_OPERATION: `${DECORATORS_PREFIX}/apiOperation`,
  API_RESPONSE: `${DECORATORS_PREFIX}/apiResponse`,
  API_PRODUCES: `${DECORATORS_PREFIX}/apiProduces`,
  API_CONSUMES: `${DECORATORS_PREFIX}/apiConsumes`,
  API_TAGS: `${DECORATORS_PREFIX}/apiUseTags`,
  API_PARAMETERS: `${DECORATORS_PREFIX}/apiParameters`,
  API_HEADERS: `${DECORATORS_PREFIX}/apiHeaders`,
  API_MODEL_PROPERTIES: `${DECORATORS_PREFIX}/apiModelProperties`,
  API_MODEL_PROPERTIES_ARRAY: `${DECORATORS_PREFIX}/apiModelPropertiesArray`,
  API_SECURITY: `${DECORATORS_PREFIX}/apiSecurity`,
  API_EXCLUDE_ENDPOINT: `${DECORATORS_PREFIX}/apiExcludeEndpoint`,
  API_EXCLUDE_CONTROLLER: `${DECORATORS_PREFIX}/apiExcludeController`,
  API_EXTRA_MODELS: `${DECORATORS_PREFIX}/apiExtraModels`,
  API_EXTENSION: `${DECORATORS_PREFIX}/apiExtension`
};
