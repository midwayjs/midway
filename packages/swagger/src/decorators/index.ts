export * from './api-body.decorator';
export * from './api-exclude-endpoint.decorator';
export * from './api-exclude-controller.decorator';
export * from './api-header.decorator';
export * from './api-operation.decorator';
export * from './api-param.decorator';
export {
  ApiProperty,
  ApiPropertyOptional,
  ApiPropertyOptions,
  ApiResponseProperty,
} from './api-property.decorator';
export * from './api-query.decorator';
export * from './api-response.decorator';
export * from './api-tags.decorator';
export * from './api-extension.decorator';
export * from './api-basic.decorator';
export * from './api-bearer.decorator';
export * from './api-cookie.decorator';
export * from './api-oauth2.decorator';
export * from './api-security.decorator';
export * from './api-extra-model.decorator';
export { getSchemaPath } from './helpers';
