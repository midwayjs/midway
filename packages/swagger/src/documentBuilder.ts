import {
  OpenAPIObject,
  ExternalDocumentationObject,
  SecurityRequirementObject,
  SecuritySchemeObject,
  ServerVariableObject,
  PathItemObject,
  SchemaObject,
  TagObject,
} from './interfaces';

export class DocumentBuilder {
  private readonly document: OpenAPIObject = {
    openapi: '3.0.1',
    info: {
      title: '',
      description: '',
      version: '1.0.0',
      contact: {},
    },
    tags: [],
    servers: [],
    components: {},
    paths: {},
  };

  public setTitle(title: string): this {
    this.document.info.title = title;
    return this;
  }

  public setDescription(description: string): this {
    this.document.info.description = description;
    return this;
  }

  public setVersion(version: string): this {
    this.document.info.version = version;
    return this;
  }

  public setTermsOfService(termsOfService: string): this {
    this.document.info.termsOfService = termsOfService;
    return this;
  }

  public setContact(name: string, url: string, email: string): this {
    this.document.info.contact = { name, url, email };
    return this;
  }

  public setLicense(name: string, url: string): this {
    this.document.info.license = { name, url };
    return this;
  }

  public addServer(
    url: string,
    description?: string,
    variables?: Record<string, ServerVariableObject>
  ): this {
    this.document.servers.push({ url, description, variables });
    return this;
  }

  public setExternalDoc(description: string, url: string): this {
    this.document.externalDocs = { description, url };
    return this;
  }

  public addPaths(paths: Record<string, PathItemObject>) {
    Object.assign(this.document.paths, paths);
    return this;
  }

  public addSchema(schema: Record<string, SchemaObject>) {
    if (!this.document.components.schemas) {
      this.document.components.schemas = {};
    }
    Object.assign(this.document.components.schemas, schema);
    return this;
  }

  public getSchema(name: string): SchemaObject {
    if (this.document.components?.schemas) {
      return this.document.components?.schemas[name] as SchemaObject;
    }
    return undefined;
  }

  public addTag(
    name: string,
    description = '',
    externalDocs?: ExternalDocumentationObject
  ): this {
    if (Array.isArray(name)) {
      const arr = name as Array<string>;
      for (const s of arr) {
        this.document.tags.push({
          name: s,
          description: '',
        });
      }
      return this;
    }
    this.document.tags.push({
      name,
      description,
      externalDocs,
    });
    return this;
  }

  public addSecurity(name: string, options: SecuritySchemeObject): this {
    this.document.components.securitySchemes = {
      ...(this.document.components.securitySchemes || {}),
      [name]: options,
    };
    return this;
  }

  public addSecurityRequirements(
    name: string | SecurityRequirementObject,
    requirements: string[] = []
  ): this {
    let securityRequirement: SecurityRequirementObject;

    if (typeof name === 'string') {
      securityRequirement = { [name as string]: requirements };
    } else {
      securityRequirement = name as SecurityRequirementObject;
    }

    this.document.security = (this.document.security || []).concat({
      ...securityRequirement,
    });
    return this;
  }

  public addBearerAuth(
    options: SecuritySchemeObject = {
      type: 'http',
    },
    name = 'bearer'
  ): this {
    this.addSecurity(name, {
      type: 'http',
      scheme: 'bearer',
      bearerFormat: 'JWT',
      ...options,
    });
    return this;
  }

  public addOAuth2(
    options: SecuritySchemeObject = {
      type: 'oauth2',
    },
    name = 'oauth2'
  ): this {
    if (!name) {
      name = 'oauth2';
    }
    this.addSecurity(name, {
      type: 'oauth2',
      flows: {
        ...options?.flows,
      },
    });
    return this;
  }

  public addApiKey(
    options: SecuritySchemeObject = {
      type: 'apiKey',
    },
    name = 'api_key'
  ): this {
    if (!name) {
      name = 'api_key';
    }
    this.addSecurity(name, {
      type: 'apiKey',
      in: 'header',
      name,
      ...options,
    });
    return this;
  }

  public addBasicAuth(
    options: SecuritySchemeObject = {
      type: 'http',
    },
    name = 'basic'
  ): this {
    if (!name) {
      name = 'basic';
    }
    this.addSecurity(name, {
      type: 'http',
      scheme: 'basic',
      ...options,
    });
    return this;
  }

  public addCookieAuth(
    cookieName = 'connect.sid',
    options: SecuritySchemeObject = {
      type: 'apiKey',
    },
    securityName = 'cookie'
  ): this {
    if (!cookieName) {
      cookieName = 'connect.sid';
    }
    if (!securityName) {
      securityName = 'cookie';
    }
    this.addSecurity(securityName, {
      type: 'apiKey',
      in: 'cookie',
      name: cookieName,
      ...options,
    });
    return this;
  }

  public sortTags(): void {
    const tags = this.document.tags;
    this.document.tags = tags.sort((a: TagObject, b: TagObject): number => {
      const s1 = a.name;
      const s2 = b.name;

      const len = s1.length > s2.length ? s2.length : s1.length;
      for (let i = 0; i < len; i++) {
        if (s1.charCodeAt(i) > s2.charCodeAt(i)) {
          return 1;
        } else if (s1.charCodeAt(i) < s2.charCodeAt(i)) {
          return -1;
        }
      }

      return 0;
    });
  }

  public build(): Omit<OpenAPIObject, 'paths'> {
    return this.document;
  }
}
