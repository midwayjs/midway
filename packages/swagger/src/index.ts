export * from './decorators';
export * from './interfaces';
export { SwaggerConfiguration as Configuration } from './configuration';
export { SwaggerExplorer } from './swaggerExplorer';
export { SwaggerMiddleware } from './swaggerMiddleware';
