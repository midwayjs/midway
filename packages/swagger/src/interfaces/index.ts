/**
 * inspired by https://github.com/metadevpro/openapi3-ts
 * @see https://github.com/OAI/OpenAPI-Specification/blob/3.0.0-rc0/versions/3.0.md
 */

export interface OpenAPIObject {
  openapi: string;
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObject;
  components?: ComponentsObject;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

export interface InfoObject {
  title: string;
  description?: string;
  termsOfService?: string;
  contact?: ContactObject;
  license?: LicenseObject;
  version: string;
}

export interface ContactObject {
  name?: string;
  url?: string;
  email?: string;
}

export interface LicenseObject {
  name: string;
  url?: string;
}

export interface ServerObject {
  url: string;
  description?: string;
  variables?: Record<string, ServerVariableObject>;
}

export interface ServerVariableObject {
  enum?: string[] | boolean[] | number[];
  default: string | boolean | number;
  description?: string;
}

export interface ComponentsObject {
  schemas?: Record<string, SchemaObject | ReferenceObject>;
  responses?: Record<string, ResponseObject | ReferenceObject>;
  parameters?: Record<string, ParameterObject | ReferenceObject>;
  examples?: Record<string, ExampleObject | ReferenceObject>;
  requestBodies?: Record<string, RequestBodyObject | ReferenceObject>;
  headers?: Record<string, HeaderObject | ReferenceObject>;
  securitySchemes?: Record<string, SecuritySchemeObject | ReferenceObject>;
  links?: Record<string, LinkObject | ReferenceObject>;
  callbacks?: Record<string, CallbackObject | ReferenceObject>;
}

export type PathsObject = Record<string, PathItemObject>;
export interface PathItemObject {
  $ref?: string;
  summary?: string;
  description?: string;
  get?: OperationObject;
  put?: OperationObject;
  post?: OperationObject;
  delete?: OperationObject;
  options?: OperationObject;
  head?: OperationObject;
  patch?: OperationObject;
  trace?: OperationObject;
  servers?: ServerObject[];
  parameters?: (ParameterObject | ReferenceObject)[];
}

export interface OperationObject {
  tags?: string[];
  summary?: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
  operationId?: string;
  parameters?: (ParameterObject | ReferenceObject)[];
  requestBody?: RequestBodyObject | ReferenceObject;
  responses: ResponsesObject;
  callbacks?: CallbacksObject;
  deprecated?: boolean;
  security?: SecurityRequirementObject[];
  servers?: ServerObject[];
}

export interface ExternalDocumentationObject {
  description?: string;
  url: string;
}

export type ParameterLocation = 'query' | 'header' | 'path' | 'cookie';
export type ParameterStyle =
  | 'matrix'
  | 'label'
  | 'form'
  | 'simple'
  | 'spaceDelimited'
  | 'pipeDelimited'
  | 'deepObject';

export interface BaseParameterObject {
  description?: string;
  required?: boolean;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  style?: ParameterStyle;
  explode?: boolean;
  allowReserved?: boolean;
  schema?: SchemaObject | ReferenceObject;
  examples?: Record<string, ExampleObject | ReferenceObject>;
  example?: any;
  content?: ContentObject;
}

export interface ParameterObject extends BaseParameterObject {
  name: string;
  in: ParameterLocation;
}

export interface RequestBodyObject {
  description?: string;
  content: ContentObject;
  required?: boolean;
}

export type ContentObject = Record<string, MediaTypeObject>;
export interface MediaTypeObject {
  schema?: SchemaObject | ReferenceObject;
  examples?: ExamplesObject;
  example?: any;
  encoding?: EncodingObject;
}

export type EncodingObject = Record<string, EncodingPropertyObject>;
export interface EncodingPropertyObject {
  contentType?: string;
  headers?: Record<string, HeaderObject | ReferenceObject>;
  style?: string;
  explode?: boolean;
  allowReserved?: boolean;
}

export interface ResponsesObject
  extends Record<string, ResponseObject | ReferenceObject | undefined> {
  default?: ResponseObject | ReferenceObject;
}

export interface ResponseObject {
  description: string;
  headers?: HeadersObject;
  content?: ContentObject;
  links?: LinksObject;
}

export type CallbacksObject = Record<string, CallbackObject | ReferenceObject>;
export type CallbackObject = Record<string, PathItemObject>;
export type HeadersObject = Record<string, HeaderObject | ReferenceObject>;

export interface ExampleObject {
  summary?: string;
  description?: string;
  value?: any;
  externalValue?: string;
}

export type LinksObject = Record<string, LinkObject | ReferenceObject>;
export interface LinkObject {
  operationRef?: string;
  operationId?: string;
  parameters?: LinkParametersObject;
  requestBody?: any | string;
  description?: string;
  server?: ServerObject;
}

export type LinkParametersObject = Record<string, any>;
export type HeaderObject = BaseParameterObject;
export interface TagObject {
  name: string;
  description?: string;
  externalDocs?: ExternalDocumentationObject;
}

export type ExamplesObject = Record<string, ExampleObject | ReferenceObject>;

export interface ReferenceObject {
  $ref: string | (() => string);
}

export interface SchemaObject {
  nullable?: boolean;
  discriminator?: DiscriminatorObject;
  readOnly?: boolean;
  writeOnly?: boolean;
  xml?: XmlObject;
  externalDocs?: ExternalDocumentationObject;
  example?: any;
  examples?: any[] | Record<string, any>;
  deprecated?: boolean;
  type?: string | (new (...args) => any) | (() => new (...args) => any);
  allOf?: (SchemaObject | ReferenceObject)[];
  oneOf?: (SchemaObject | ReferenceObject)[];
  anyOf?: (SchemaObject | ReferenceObject)[];
  not?: SchemaObject | ReferenceObject;
  items?: SchemaObject | ReferenceObject;
  properties?: Record<string, SchemaObject | ReferenceObject>;
  additionalProperties?: SchemaObject | ReferenceObject | boolean;
  patternProperties?: SchemaObject | ReferenceObject | any;
  description?: string;
  format?: string;
  default?: any;
  title?: string;
  multipleOf?: number;
  maximum?: number;
  exclusiveMaximum?: boolean;
  minimum?: number;
  exclusiveMinimum?: boolean;
  maxLength?: number;
  minLength?: number;
  pattern?: string;
  maxItems?: number;
  minItems?: number;
  uniqueItems?: boolean;
  maxProperties?: number;
  minProperties?: number;
  required?: string[];
  enum?: any[];
}

export type SchemasObject = Record<string, SchemaObject>;

export interface DiscriminatorObject {
  propertyName: string;
  mapping?: Record<string, string>;
}

export interface XmlObject {
  name?: string;
  namespace?: string;
  prefix?: string;
  attribute?: boolean;
  wrapped?: boolean;
}

export type SecuritySchemeType = 'apiKey' | 'http' | 'oauth2' | 'openIdConnect';

export interface SecuritySchemeObject {
  type: SecuritySchemeType;
  description?: string;
  name?: string;
  in?: string;
  scheme?: string;
  bearerFormat?: string;
  flows?: OAuthFlowsObject;
  openIdConnectUrl?: string;
}

export interface OAuthFlowsObject {
  implicit?: OAuthFlowObject;
  password?: OAuthFlowObject;
  clientCredentials?: OAuthFlowObject;
  authorizationCode?: OAuthFlowObject;
}

export interface OAuthFlowObject {
  authorizationUrl?: string;
  tokenUrl?: string;
  refreshUrl?: string;
  scopes: ScopesObject;
}

export type ScopesObject = Record<string, any>;
export type SecurityRequirementObject = Record<string, string[]>;

/**
 * 非 open api spec
 */
export type SwaggerEnumType =
  | string[]
  | number[]
  | (string | number)[]
  | Record<number, string>;

export interface Type<T = any> extends Function {
  new (...args: any[]): T;
}

export interface SchemaObjectMetadata
  extends Omit<SchemaObject, 'type' | 'required'> {
  type?: Type<unknown> | [Type] | string | Record<string, any>;
  isArray?: boolean;
  required?: boolean;
  name?: string;
  enumName?: string;
}

export type AuthType =
  | 'basic'
  | 'bearer'
  | 'cookie'
  | 'oauth2'
  | 'apikey'
  | 'custom';

/**
 * 继承自 https://swagger.io/specification/#security-scheme-object
 */
export interface AuthOptions extends Omit<SecuritySchemeObject, 'type'> {
  /**
   * 验权类型
   * basic  => http basic 验证
   * bearer => http jwt 验证
   * cookie => cookie 方式验证
   * oauth2 => 使用 oauth2
   * apikey => apiKey
   * custom => 自定义方式
   */
  authType: AuthType;
  /**
   * https://swagger.io/specification/#security-scheme-object type 字段
   */
  type?: SecuritySchemeType;
  /**
   * authType = cookie 时可以修改，通过 ApiCookie 装饰器关联的名称
   */
  securityName?: string;
  /**
   * authType = cookie 时可以修改，cookie 的名称
   */
  cookieName?: string;
}
/**
 * see https://swagger.io/specification/
 */
export interface SwaggerOptions {
  /**
   * 默认值: My Project
   * https://swagger.io/specification/#info-object title 字段
   */
  title?: string;
  /**
   * 默认值: This is a swagger-ui for midwayjs project
   * https://swagger.io/specification/#info-object description 字段
   */
  description?: string;
  /**
   * 默认值: 1.0.0
   * https://swagger.io/specification/#info-object version 字段
   */
  version?: string;
  /**
   * https://swagger.io/specification/#info-object contact 字段
   */
  contact?: ContactObject;
  /**
   * https://swagger.io/specification/#info-object license 字段
   */
  license?: LicenseObject;
  /**
   * https://swagger.io/specification/#info-object termsOfService 字段
   */
  termsOfService?: string;
  /**
   * https://swagger.io/specification/#openapi-object externalDocs 字段
   */
  externalDocs?: ExternalDocumentationObject;
  /**
   * https://swagger.io/specification/#openapi-object servers 字段
   */
  servers?: Array<ServerObject>;
  /**
   * https://swagger.io/specification/#openapi-object tags 字段
   */
  tags?: Array<TagObject>;
  /**
   * 可以参考 https://swagger.io/specification/#security-scheme-object
   */
  auth?: AuthOptions | AuthOptions[];
  /**
   * 默认值: /swagger-ui
   * 访问 swagger ui 的路径
   */
  swaggerPath?: string;
  /**
   * 对路由 tag 进行 ascii 排序
   * 可以使用 1-xxx、2-xxx、3-xxx 来定义 tag
   */
  tagSortable?: boolean;
  /**
   * UI 展示中需要用到的配置
   * 可以参考 https://github.com/swagger-api/swagger-ui/blob/master/docs/usage/configuration.md#display
   */
  displayOptions?: {
    deepLinking?: boolean;
    displayOperationId?: boolean;
    defaultModelsExpandDepth?: number;
    defaultModelExpandDepth?: number;
    defaultModelRendering?: 'example' | 'model';
    displayRequestDuration?: boolean;
    docExpansion?: 'list' | 'full' | 'none';
    filter?: boolean | string;
    maxDisplayedTags?: number;
    showExtensions?: boolean;
    showCommonExtensions?: boolean;
    useUnsafeMarkdown?: boolean;
    tryItOutEnabled?: boolean;
  };
}
