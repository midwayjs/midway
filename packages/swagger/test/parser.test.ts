import { ApiExtraModel, ApiProperty, ApiResponse, getSchemaPath, SwaggerExplorer, Type } from '../src';
import { Controller, Post } from '@midwayjs/core';

class CustomSwaggerExplorer extends SwaggerExplorer {
  generatePath(target: Type) {
    return super.generatePath(target);
  }

  parseClzz(clz) {
    return super.parseClzz(clz);
  }
}

describe('/test/parser.test.ts', function () {
  it('should fix issue#2286 with array example', function () {
    class Catd {
      @ApiProperty({ example: 'Kitty', description: 'The name of the Cat'})
      named: string;

      @ApiProperty({ example: 1, description: 'The age of the Cat' })
      aged: number;

      @ApiProperty({
        example: 'Maine Coon',
        description: 'The breed of the Cat',
      })
      breedd: string;
    }


    @ApiExtraModel(Catd)
    class Cat {
      @ApiProperty({ example: 'Kitty', description: 'The name of the Cat'})
      name: string;

      @ApiProperty({
        description: 'The breed of the Cat',
        required: false,
        type: 'array',
        items: {
          $ref: getSchemaPath(Catd)
        },
      })
      catds: Catd[];
    }

    @Controller('/api', { middleware: []})
    class CatController {
      @ApiResponse({
        status: 200,
        description: 'The found record',
        type: Cat,
      })
      @Post('/update_user')
      async updateUser() {
        // ...
      }
    }

    const explorer = new CustomSwaggerExplorer();
    explorer.generatePath(CatController);
    expect(explorer.getData()).toMatchSnapshot();
  });

  it('should test simple case with example', function () {
    class Cat {
      @ApiProperty({ example: 'Kitty', description: 'The name of the Cat'})
      name: string;
    }

    @Controller('/api', { middleware: []})
    class CatController {
      @ApiResponse({
        status: 200,
        description: 'The found record',
        type: Cat,
      })
      @Post('/update_user')
      async updateUser() {
        // ...
      }
    }

    const explorer = new CustomSwaggerExplorer();
    explorer.generatePath(CatController);
    expect(explorer.getData()).toMatchSnapshot();
  });

  it('should test ref path generate', function () {
    class NotificationDTO {
      @ApiProperty({ example: '通知标题', description: 'title'})
      title: string;

      @ApiProperty({ example: '1', description: '这是 id'})
      id: number;
    }

    @ApiExtraModel(NotificationDTO)
    class NotificationPageListDTO {
      @ApiProperty({ description: '列表数据', type: 'array', items: { $ref: getSchemaPath(NotificationDTO) } })
      rows: NotificationDTO[]

      @ApiProperty({ example: '999', description: '通知数量'})
      count: number;
    }
    @ApiExtraModel(NotificationPageListDTO)
    class UserDTO {
      @ApiProperty({ example: 'Kitty', description: 'The name of the user'})
      name: string;

      @ApiProperty({ example: '1', description: 'The uid of the user'})
      uid: number;

      @ApiProperty({ description: 'The uid of the user', type: NotificationPageListDTO})
      data: NotificationPageListDTO
    }

    @Controller('/api', { middleware: []})
    class APIController {
      @ApiResponse({
        status: 200,
        description: 'The found record',
        type: UserDTO,
      })
      @Post('/update_user')
      async updateUser() {
        // ...
      }
    }

    const explorer = new CustomSwaggerExplorer();
    explorer.generatePath(APIController);
    expect(explorer.getData()).toMatchSnapshot();
  });

  it('should test specify type', function () {
    class NotificationDTO {
      @ApiProperty({ type: 'string' })
      id: number;
    }

    const explorer = new CustomSwaggerExplorer();
    expect(explorer.parseClzz(NotificationDTO)).toMatchSnapshot();
  });


  it('should parse ref schema with function', function () {
    class Photo {
      @ApiProperty()
      id: number;

      @ApiProperty()
      name: string;

      @ApiProperty({
        type: 'array',
        items: {
          $ref: () => getSchemaPath(Album),
        }
      })
      albums: Album[];
    }
    class Album {
      @ApiProperty()
      id: number;

      @ApiProperty()
      name: string;

      @ApiProperty({
        type: 'array',
        items: {
          $ref: () => getSchemaPath(Photo),
        }
      })
      photos: Photo[];
    }

    const explorer = new CustomSwaggerExplorer();
    expect(explorer.parseClzz(Photo)).toMatchSnapshot();
    expect(explorer.parseClzz(Album)).toMatchSnapshot();
  });

  it('should parse type with function', function () {
    class Photo {
      @ApiProperty()
      id: number;

      @ApiProperty()
      name: string;

      @ApiProperty({
        type: () => {
          return Album;
        }
      })
      album: any;
    }
    class Album {
      @ApiProperty()
      id: number;

      @ApiProperty()
      name: string;

      @ApiProperty({
        type: Photo,
      })
      photo: Photo;
    }

    const explorer = new CustomSwaggerExplorer();
    expect(explorer.parseClzz(Photo)).toMatchSnapshot();
    expect(explorer.parseClzz(Album)).toMatchSnapshot();
  });

  it('should test multi-type in property', function () {
    class Album {
      @ApiProperty()
      id: number;

      @ApiProperty()
      name: string;
    }
    class Photo {
      @ApiProperty({
        oneOf: [
          { type: 'string' },
          {
            type: 'array',
            items: {
              type: 'string'
            }
          }
        ]
      })
      name: string | string[];

      @ApiProperty({
        oneOf: [
          { type: Album},
          {
            type: 'array',
            items: {
              type: () => Album,
            }
          }
        ]
      })
      album: Album | Album[];
    }
    const explorer = new CustomSwaggerExplorer();
    expect(explorer.parseClzz(Photo)).toMatchSnapshot();
  });

  it('should parse base type', function () {
    class Cat {
      @ApiProperty({
        type: [String],
        example: ['1'],
        description: 'The name of the Catage',
        nullable: true,
        uniqueItems: true,
      })
      breeds: string[];
    }

    const explorer = new CustomSwaggerExplorer();
    expect(explorer.parseClzz(Cat)).toMatchSnapshot();
  });
});
