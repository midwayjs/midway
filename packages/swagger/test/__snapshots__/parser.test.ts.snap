// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/parser.test.ts should fix issue#2286 with array example 1`] = `
{
  "components": {
    "schemas": {
      "Cat": {
        "properties": {
          "catds": {
            "description": "The breed of the Cat",
            "items": {
              "$ref": "#/components/schemas/Catd",
            },
            "type": "array",
          },
          "name": {
            "description": "The name of the Cat",
            "example": "Kitty",
            "format": undefined,
            "type": "string",
          },
        },
        "type": "object",
      },
      "Catd": {
        "properties": {
          "aged": {
            "description": "The age of the Cat",
            "example": 1,
            "format": undefined,
            "type": "number",
          },
          "breedd": {
            "description": "The breed of the Cat",
            "example": "Maine Coon",
            "format": undefined,
            "type": "string",
          },
          "named": {
            "description": "The name of the Cat",
            "example": "Kitty",
            "format": undefined,
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "contact": {},
    "description": "",
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": {
    "/api/update_user": {
      "post": {
        "description": undefined,
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Cat",
                },
              },
            },
            "description": "The found record",
          },
        },
        "summary": undefined,
        "tags": [
          "api",
        ],
      },
    },
  },
  "servers": [],
  "tags": [
    {
      "description": "api",
      "externalDocs": undefined,
      "name": "api",
    },
  ],
}
`;

exports[`/test/parser.test.ts should parse base type 1`] = `
{
  "properties": {
    "breeds": {
      "description": "The name of the Catage",
      "example": [
        "1",
      ],
      "items": {
        "type": "string",
      },
      "nullable": true,
      "type": "array",
      "uniqueItems": true,
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should parse ref schema with function 1`] = `
{
  "properties": {
    "albums": {
      "items": {
        "$ref": "#/components/schemas/Album",
      },
      "type": "array",
    },
    "id": {
      "format": undefined,
      "type": "number",
    },
    "name": {
      "format": undefined,
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should parse ref schema with function 2`] = `
{
  "properties": {
    "id": {
      "format": undefined,
      "type": "number",
    },
    "name": {
      "format": undefined,
      "type": "string",
    },
    "photos": {
      "items": {
        "$ref": "#/components/schemas/Photo",
      },
      "type": "array",
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should parse type with function 1`] = `
{
  "properties": {
    "album": {
      "$ref": "#/components/schemas/Album",
    },
    "id": {
      "format": undefined,
      "type": "number",
    },
    "name": {
      "format": undefined,
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should parse type with function 2`] = `
{
  "properties": {
    "id": {
      "format": undefined,
      "type": "number",
    },
    "name": {
      "format": undefined,
      "type": "string",
    },
    "photo": {
      "$ref": "#/components/schemas/Photo",
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should test multi-type in property 1`] = `
{
  "properties": {
    "album": {
      "oneOf": [
        {
          "$ref": "#/components/schemas/Album",
        },
        {
          "items": {
            "$ref": "#/components/schemas/Album",
          },
          "type": "array",
        },
      ],
    },
    "name": {
      "oneOf": [
        {
          "format": undefined,
          "type": "string",
        },
        {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      ],
    },
  },
  "type": "object",
}
`;

exports[`/test/parser.test.ts should test ref path generate 1`] = `
{
  "components": {
    "schemas": {
      "NotificationDTO": {
        "properties": {
          "id": {
            "description": "这是 id",
            "example": "1",
            "format": undefined,
            "type": "number",
          },
          "title": {
            "description": "title",
            "example": "通知标题",
            "format": undefined,
            "type": "string",
          },
        },
        "type": "object",
      },
      "NotificationPageListDTO": {
        "properties": {
          "count": {
            "description": "通知数量",
            "example": "999",
            "format": undefined,
            "type": "number",
          },
          "rows": {
            "description": "列表数据",
            "items": {
              "$ref": "#/components/schemas/NotificationDTO",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "UserDTO": {
        "properties": {
          "data": {
            "$ref": "#/components/schemas/NotificationPageListDTO",
            "description": "The uid of the user",
          },
          "name": {
            "description": "The name of the user",
            "example": "Kitty",
            "format": undefined,
            "type": "string",
          },
          "uid": {
            "description": "The uid of the user",
            "example": "1",
            "format": undefined,
            "type": "number",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "contact": {},
    "description": "",
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": {
    "/api/update_user": {
      "post": {
        "description": undefined,
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserDTO",
                },
              },
            },
            "description": "The found record",
          },
        },
        "summary": undefined,
        "tags": [
          "api",
        ],
      },
    },
  },
  "servers": [],
  "tags": [
    {
      "description": "api",
      "externalDocs": undefined,
      "name": "api",
    },
  ],
}
`;

exports[`/test/parser.test.ts should test simple case with example 1`] = `
{
  "components": {
    "schemas": {
      "Cat": {
        "properties": {
          "name": {
            "description": "The name of the Cat",
            "example": "Kitty",
            "format": undefined,
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "contact": {},
    "description": "",
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": {
    "/api/update_user": {
      "post": {
        "description": undefined,
        "parameters": [],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Cat",
                },
              },
            },
            "description": "The found record",
          },
        },
        "summary": undefined,
        "tags": [
          "api",
        ],
      },
    },
  },
  "servers": [],
  "tags": [
    {
      "description": "api",
      "externalDocs": undefined,
      "name": "api",
    },
  ],
}
`;

exports[`/test/parser.test.ts should test specify type 1`] = `
{
  "properties": {
    "id": {
      "format": undefined,
      "type": "string",
    },
  },
  "type": "object",
}
`;
